File: src/main/java/me/sd_master92/customvoting/subjects/CustomVote.java
package me.sd_master92.customvoting.subjects;

import com.vexsoftware.votifier.model.Vote;
import com.vexsoftware.votifier.model.VotifierEvent;
import me.sd_master92.customfile.PlayerFile;
import me.sd_master92.customvoting.Main;
import me.sd_master92.customvoting.VoteFile;
import me.sd_master92.customvoting.constants.Data;
import me.sd_master92.customvoting.constants.Messages;
import me.sd_master92.customvoting.constants.Settings;
import net.md_5.bungee.api.ChatMessageType;
import net.md_5.bungee.api.chat.TextComponent;
import net.milkbowl.vault.economy.Economy;
import org.bukkit.*;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Firework;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.FireworkMeta;
import org.bukkit.scheduler.BukkitRunnable;

import java.text.DecimalFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class CustomVote extends Vote
{
    private static boolean isAwaitingBroadcast = false;

    private final Main plugin;

    public CustomVote(Main plugin, Vote vote)
    {
        super();
        this.plugin = plugin;
        setUsername(vote.getUsername());
        setServiceName(vote.getServiceName());
        setAddress(vote.getAddress());
        setTimeStamp(vote.getTimeStamp());

        forwardVote();
    }

    public static void shootFirework(Main plugin, Location loc)
    {
        if (plugin.getConfig().getBoolean(Settings.FIREWORK))
        {
            World world = loc.getWorld();
            if (world != null)
            {
                Firework firework = (Firework) world.spawnEntity(loc, EntityType.FIREWORK);
                FireworkMeta fireworkMeta = firework.getFireworkMeta();
                Random random = new Random();
                Color[] colors = {Color.AQUA, Color.BLUE, Color.FUCHSIA, Color.GREEN, Color.LIME, Color.MAROON,
                        Color.NAVY,
                        Color.OLIVE, Color.ORANGE, Color.PURPLE, Color.RED, Color.TEAL,};
                FireworkEffect.Type[] fireworkEffects = {FireworkEffect.Type.BALL, FireworkEffect.Type.BALL_LARGE,
                        FireworkEffect.Type.BURST, FireworkEffect.Type.STAR};
                FireworkEffect effect =
                        FireworkEffect.builder()
                                .flicker(random.nextBoolean())
                                .withColor(colors[random.nextInt(colors.length)])
                                .withFade(colors[random.nextInt(colors.length)])
                                .with(fireworkEffects[random.nextInt(fireworkEffects.length)])
                                .trail(random.nextBoolean()).build();
                fireworkMeta.addEffect(effect);
                fireworkMeta.setPower(random.nextInt(2) + 1);
                firework.setFireworkMeta(fireworkMeta);
            }
        }
    }

    public static void create(Main plugin, String name, String service)
    {
        Vote vote = new Vote();
        vote.setUsername(name);
        vote.setServiceName(service);
        vote.setAddress("0.0.0.0");
        Date date = new Date();
        vote.setTimeStamp(String.valueOf(date.getTime()));
        plugin.getServer().getPluginManager().callEvent(new VotifierEvent(vote));
    }

    private void forwardVote()
    {
        Player player = Bukkit.getPlayer(getUsername());
        if (player == null)
        {
            queue();
        } else
        {
            new VoteFile(player.getUniqueId().toString(), plugin).addVote(true);
            broadcast();
            shootFirework(plugin, player.getLocation());
            giveRewards(player);
            if (plugin.getConfig().getBoolean(Settings.VOTE_PARTY))
            {
                subtractVotesUntilVoteParty();
            }
        }
    }

    private void queue()
    {
        PlayerFile playerFile = PlayerFile.getByName(getUsername(), plugin);
        if (playerFile != null)
        {
            new VoteFile(playerFile.getUuid(), plugin).addQueue(getServiceName());
        }
    }

    private void broadcast()
    {
        HashMap<String, String> placeholders = new HashMap<>();
        placeholders.put("%PLAYER%", getUsername());
        placeholders.put("%SERVICE%", getServiceName());
        String message = Messages.VOTE_BROADCAST.getMessage(plugin, placeholders);
        plugin.getServer().broadcastMessage(message);
    }

    private void subtractVotesUntilVoteParty()
    {
        if (plugin.getData().getLocations(Data.VOTE_PARTY).size() > 0)
        {
            int votesRequired = plugin.getConfig().getNumber(Settings.VOTES_REQUIRED_FOR_VOTE_PARTY);
            int votesUntil = votesRequired - plugin.getData().getNumber(Data.CURRENT_VOTES);
            if (votesUntil <= 1)
            {
                plugin.getData().setNumber(Data.CURRENT_VOTES, 0);
                new BukkitRunnable()
                {
                    @Override
                    public void run()
                    {
                        new VoteParty(plugin).start();
                    }
                }.runTaskLater(plugin, 40);
            } else
            {
                plugin.getData().addNumber(Data.CURRENT_VOTES);
                if (!isAwaitingBroadcast)
                {
                    isAwaitingBroadcast = true;
                    new BukkitRunnable()
                    {
                        @Override
                        public void run()
                        {
                            int updatedVotesUntil = votesRequired - plugin.getData().getNumber(Data.CURRENT_VOTES);
                            if (updatedVotesUntil != votesRequired)
                            {
                                Map<String, String> placeholders = new HashMap<>();
                                placeholders.put("%VOTES%", "" + updatedVotesUntil);
                                placeholders.put("%s%", updatedVotesUntil == 1 ? "" : "s");
                                plugin.getServer().broadcastMessage(Messages.VOTE_PARTY_UNTIL.getMessage(plugin, placeholders));
                            }
                            isAwaitingBroadcast = false;
                        }
                    }.runTaskLater(plugin, 40);
                }
            }
        }
    }

    private void giveRewards(Player player)
    {
        giveItems(player);
        giveLuckyReward(player);
        executeCommands(player);
        String rewardMessage = "";
        double money = giveMoney(player);
        if (Main.economy != null && money > 0)
        {
            Map<String, String> placeholders = new HashMap<>();
            placeholders.put("%MONEY%", new DecimalFormat("#.##").format(money));
            rewardMessage += Messages.VOTE_REWARD_MONEY.getMessage(plugin, placeholders);
        }
        int xp = giveExperience(player);
        if (xp > 0)
        {
            Map<String, String> placeholders = new HashMap<>();
            placeholders.put("%XP%", "" + xp);
            placeholders.put("%s%", xp == 1 ? "" : "s");
            rewardMessage += rewardMessage.isEmpty() ? "" :
                    Messages.VOTE_REWARD_DIVIDER.getMessage(plugin);
            rewardMessage += Messages.VOTE_REWARD_XP.getMessage(plugin, placeholders);
        }
        String message = rewardMessage;
        if (!message.isEmpty())
        {
            new BukkitRunnable()
            {
                int i = 40;

                @Override
                public void run()
                {
                    player.spigot().sendMessage(ChatMessageType.ACTION_BAR,
                            new TextComponent(Messages.VOTE_REWARD_PREFIX.getMessage(plugin) + message));
                    if (i == 0)
                    {
                        cancel();
                    }
                    i--;
                }
            }.runTaskTimer(plugin, 0, 1);
        }
    }

    private void giveItems(Player player)
    {
        for (ItemStack reward : plugin.getData().getItems(Data.ITEM_REWARDS))
        {
            for (ItemStack item : player.getInventory().addItem(reward).values())
            {
                player.getWorld().dropItemNaturally(player.getLocation(), item);
            }
        }
    }

    private double giveMoney(Player player)
    {
        Economy economy = Main.economy;
        if (economy != null && economy.hasAccount(player))
        {
            double amount = plugin.getConfig().getDouble(Settings.VOTE_REWARD_MONEY);
            economy.depositPlayer(player, amount);
            return amount;
        }
        return 0;
    }

    private int giveExperience(Player player)
    {
        int amount = plugin.getConfig().getNumber(Settings.VOTE_REWARD_EXPERIENCE);
        player.setLevel(player.getLevel() + amount);
        return amount;
    }

    private void giveLuckyReward(Player player)
    {
        Random random = new Random();
        int i = random.nextInt(100);
        if (i < plugin.getConfig().getNumber(Settings.LUCKY_VOTE_CHANCE))
        {
            ItemStack[] luckyRewards = plugin.getData().getItems(Data.LUCKY_REWARDS);
            if (luckyRewards.length > 0)
            {
                i = random.nextInt(luckyRewards.length);
                for (ItemStack item : player.getInventory().addItem(luckyRewards[i]).values())
                {
                    player.getWorld().dropItemNaturally(player.getLocation(), item);
                }
                player.sendMessage(Messages.VOTE_LUCKY.getMessage(plugin));
            }
        }
    }

    private void executeCommands(Player player)
    {
        for (String command : plugin.getData().getStringList(Data.VOTE_COMMANDS))
        {
            plugin.getServer().dispatchCommand(plugin.getServer().getConsoleSender(), command.replace("%PLAYER%",
                    player.getName()));
        }
    }
}

File: src/main/kotlin/me/sd_master92/customvoting/subjects/CustomVote.kt
package me.sd_master92.customvoting.subjects

import com.github.shynixn.mccoroutine.bukkit.launch
import com.vexsoftware.votifier.model.Vote
import me.sd_master92.core.appendWhenTrue
import me.sd_master92.core.errorLog
import me.sd_master92.core.infoLog
import me.sd_master92.core.tasks.TaskTimer
import me.sd_master92.customvoting.*
import me.sd_master92.customvoting.constants.enumerations.*
import me.sd_master92.customvoting.constants.interfaces.Voter
import me.sd_master92.customvoting.constants.models.VoteSiteUUID
import me.sd_master92.customvoting.helpers.ParticleHelper
import me.sd_master92.customvoting.subjects.voteparty.VoteParty
import me.sd_master92.customvoting.subjects.voteparty.VotePartyChest
import org.bukkit.entity.Player
import java.text.DecimalFormat
import java.time.LocalDate
import java.util.*

class CustomVote(
    private val plugin: CV,
    vote: Vote,
    private var queued: Boolean = false,
    private val silent: Boolean = false
) : Vote(vote)
{
    private var previousLast: Long = 0
    private var votes = 0

    private suspend fun forwardVote()
    {
        val player = username.getPlayer(plugin)
        if (player == null)
        {
            plugin.infoLog(PMessage.QUEUE_MESSAGE_ADD.toString())
            register(true)
        } else if (plugin.config.getStringList(Setting.DISABLED_WORLDS.path).contains(player.world.name))
        {
            if (!plugin.config.getBoolean(Setting.DISABLED_MESSAGE_DISABLED_WORLD.path))
            {
                player.sendText(plugin, Message.DISABLED_WORLD)
            }
            register(true)
        } else
        {
            if (!silent)
            {
                broadcast(player)
            }
            val voter = Voter.get(plugin, player)
            previousLast = voter.getLast()
            votes = voter.getVotes()
            voter.addVote()
            ParticleHelper.shootFirework(plugin, player.location)
            giveRewards(player, player.hasPowerRewards(plugin))
            if (plugin.config.getBoolean(Setting.VOTE_PARTY_ENABLED.path))
            {
                subtractVotesUntilVoteParty()
            }
            register()
        }
    }

    private suspend fun register(queue: Boolean = false)
    {
        if (!queued)
        {
            val voter = Voter.getByName(plugin, username)
            if (voter != null)
            {
                voter.addHistory(VoteSiteUUID(serviceName), queue)
            } else
            {
                plugin.errorLog(PMessage.PLAYER_ERROR_NOT_EXIST_X.with(username))
            }
        }
    }

    private suspend fun broadcast(player: Player)
    {
        if (!plugin.config.getBoolean(Setting.DISABLED_BROADCAST_VOTE.path))
        {
            if (plugin.config.getBoolean(Setting.FIRST_VOTE_BROADCAST_ONLY.path))
            {
                val last = Voter.get(plugin, player).getLast()
                val date1 = Calendar.getInstance()
                date1.time = Date(last)
                val date2 = Calendar.getInstance()
                date2.time = Date()
                if (date1.get(Calendar.DAY_OF_YEAR) == date2.get(Calendar.DAY_OF_YEAR) &&
                    date1.get(Calendar.YEAR) == date2.get(Calendar.YEAR)
                )
                {
                    return
                }
            }

            val placeholders = HashMap<String, String>()
            placeholders["%PLAYER%"] = username
            placeholders["%SERVICE%"] = serviceName
            plugin.broadcastText(Message.VOTE_BROADCAST_SINGLE, placeholders)
        }
    }

    private fun subtractVotesUntilVoteParty()
    {
        if (VotePartyChest.getAll(plugin).isNotEmpty())
        {
            val votesRequired = plugin.config.getNumber(Setting.VOTES_REQUIRED_FOR_VOTE_PARTY.path)
            val votesUntil = votesRequired - plugin.data.getNumber(Data.CURRENT_VOTES.path)
            if (votesUntil <= 1)
            {
                plugin.data.setNumber(Data.CURRENT_VOTES.path, 0)
                TaskTimer.delay(plugin, 40) { VoteParty(plugin).start() }.run()
            } else
            {
                plugin.data.addNumber(Data.CURRENT_VOTES.path)
                if (!isAwaitingBroadcast)
                {
                    isAwaitingBroadcast = true
                    TaskTimer.delay(plugin, 40)
                    {
                        val updatedVotesUntil = votesRequired - plugin.data.getNumber(Data.CURRENT_VOTES.path)
                        if (updatedVotesUntil != votesRequired && !plugin.config.getBoolean(Setting.DISABLED_BROADCAST_VOTE_PARTY_UNTIL.path))
                        {
                            val placeholders = HashMap<String, String>()
                            placeholders["%VOTES%"] = "$updatedVotesUntil"
                            placeholders["%s%"] = if (updatedVotesUntil == 1) "" else "s"
                            plugin.broadcastText(Message.VOTE_PARTY_UNTIL, placeholders)
                        }
                        isAwaitingBroadcast = false
                    }.run()
                }
            }
        }
    }

    private suspend fun giveRewards(player: Player, power: Boolean)
    {
        val dayOfWeek = LocalDate.now().getDayOfWeek()
        val doubleRewards = dayOfWeek.isDoubleRewards(plugin, power)

        giveItems(player, power, doubleRewards)
        executeCommands(player, Data.VOTE_COMMANDS.path.appendWhenTrue(power, Data.POWER_REWARDS))
        var rewardMessage = ""
        val money = giveMoney(player, power, doubleRewards)
        if (CV.ECONOMY != null && money > 0)
        {
            val placeholders = HashMap<String, String>()
            placeholders["%MONEY%"] = DecimalFormat("#.##").format(money)
            rewardMessage += Message.VOTE_REWARD_MONEY.getMessage(plugin, placeholders)
        }
        val xp = giveExperience(player, power, doubleRewards)
        if (xp > 0)
        {
            val placeholders = HashMap<String, String>()
            placeholders["%XP%"] = "$xp"
            placeholders["%s%"] = if (xp == 1) "" else "s"
            rewardMessage += if (rewardMessage.isEmpty()) "" else Message.VOTE_REWARD_DIVIDER.getMessage(plugin)
            rewardMessage += Message.VOTE_REWARD_XP.getMessage(plugin, placeholders)
        }
        giveLuckyReward(player)
        giveMilestoneRewards(player)
        giveStreakRewards(player)

        val message = rewardMessage
        if (message.isNotEmpty())
        {
            TaskTimer.repeat(plugin, 1, 0, 40)
            {
                player.sendActionBar(Message.VOTE_REWARD_PREFIX.getMessage(plugin) + message)
            }.run()
        }

        if (doubleRewards)
        {
            val placeholders = HashMap<String, String>()
            placeholders["%PLAYER%"] = username
            placeholders["%DAY%"] = dayOfWeek.getDisplayName(plugin).lowercase()
            plugin.broadcastText(Message.DOUBLE_REWARDS_RECEIVED, placeholders)
        }
    }

    private fun giveItems(player: Player, power: Boolean, doubleRewards: Boolean)
    {
        val path = Data.ITEM_REWARDS.path.appendWhenTrue(power, Data.POWER_REWARDS)
        val typePath = Setting.ITEM_REWARD_TYPE.path.appendWhenTrue(power, Setting.POWER_REWARDS)
        val random = plugin.config.getNumber(typePath) != ItemRewardType.ALL_ITEMS.ordinal
        val items = plugin.data.getItems(path)
        player.addToInventoryOrDrop(items, random)
        if (doubleRewards)
        {
            player.addToInventoryOrDrop(items, random)
        }
    }

    private fun giveMoney(player: Player, power: Boolean, doubleRewards: Boolean): Double
    {
        val economy = CV.ECONOMY
        if (economy != null && economy.hasAccount(player))
        {
            val path = Setting.VOTE_REWARD_MONEY.path.appendWhenTrue(power, Setting.POWER_REWARDS)
            var amount = plugin.config.getDouble(path)

            if (doubleRewards)
            {
                amount *= 2
            }

            economy.depositPlayer(player, amount)
            return amount
        }
        return 0.0
    }

    private fun giveExperience(player: Player, power: Boolean, doubleRewards: Boolean): Int
    {
        val path = Setting.VOTE_REWARD_EXPERIENCE.path.appendWhenTrue(power, Setting.POWER_REWARDS)
        var amount = plugin.config.getNumber(path)

        if (doubleRewards)
        {
            amount *= 2
        }

        player.level += amount
        return amount
    }

    private fun giveLuckyReward(player: Player)
    {
        if (plugin.config.getBoolean(Setting.LUCKY_VOTE.path) && Random().nextInt(100) < plugin.config.getNumber(Setting.LUCKY_VOTE_CHANCE.path))
        {
            val luckyRewards = plugin.data.getItems(Data.LUCKY_REWARDS.path)
            if (luckyRewards.isNotEmpty())
            {
                player.addToInventoryOrDrop(luckyRewards, true)
                player.sendMessage(Message.VOTE_LUCKY.getMessage(plugin))
            }
            executeCommands(player, Data.LUCKY_COMMANDS.path)
        }
    }

    private fun executeCommands(player: Player, dataPath: String)
    {
        val commands = plugin.data.getStringList(dataPath)
        for (command in commands)
        {
            plugin.runCommand(command.replace("%PLAYER%", player.name).withPlaceholders(player))
        }
    }

    private suspend fun giveMilestoneRewards(player: Player)
    {
        val votes = Voter.get(plugin, player).getVotes()
        if (plugin.data.contains(Data.MILESTONES.path + ".$votes"))
        {
            if (!plugin.config.getBoolean(Setting.DISABLED_BROADCAST_MILESTONE.path))
            {
                val placeholders = HashMap<String, String>()
                placeholders["%PLAYER%"] = player.name
                placeholders["%MILESTONE%"] = "$votes"
                player.sendText(plugin, Message.MILESTONE_REACHED, placeholders)
            }

            val permissions = plugin.data.getStringList(Data.MILESTONES.path + ".$votes.permissions")
            if (CV.PERMISSION != null)
            {
                for (permission in permissions)
                {
                    CV.PERMISSION!!.playerAdd(null, player, permission)
                }
            }
            val commands = plugin.data.getStringList(Data.MILESTONES.path + ".$votes.commands")
            for (command in commands)
            {
                plugin.runCommand(command.replace("%PLAYER%", player.name).withPlaceholders(player))
            }
            player.addToInventoryOrDrop(plugin.data.getItems("${Data.MILESTONES}.$votes.${Data.ITEM_REWARDS}"))
        }
    }

    private suspend fun giveStreakRewards(player: Player)
    {
        val voter = Voter.get(plugin, player)
        val streak = voter.getStreakDaily()
        if (plugin.data.contains(Data.STREAKS.path + ".$streak"))
        {
            if (previousLast.dayDifferenceToday() == 1 || votes == 0)
            {
                if (!plugin.config.getBoolean(Setting.DISABLED_BROADCAST_STREAK.path))
                {
                    val placeholders = HashMap<String, String>()
                    placeholders["%PLAYER%"] = player.name
                    placeholders["%STREAK%"] = "$streak"
                    player.sendText(plugin, Message.STREAK_REACHED, placeholders)
                }

                val permissions = plugin.data.getStringList(Data.STREAKS.path + ".$streak.permissions")
                if (CV.PERMISSION != null)
                {
                    for (permission in permissions)
                    {
                        CV.PERMISSION!!.playerAdd(null, player, permission)
                    }
                }
                val commands = plugin.data.getStringList(Data.STREAKS.path + ".$streak.commands")
                for (command in commands)
                {
                    plugin.runCommand(command.replace("%PLAYER%", player.name).withPlaceholders(player))
                }
                player.addToInventoryOrDrop(plugin.data.getItems("${Data.STREAKS}.$streak.${Data.ITEM_REWARDS}"))
            }
        }
    }

    companion object
    {
        private var isAwaitingBroadcast = false

        fun create(
            plugin: CV,
            name: String,
            service: String,
            queued: Boolean = false,
            silent: Boolean = false
        )
        {
            val vote = Vote(service, name, "0.0.0.0", Date().time.toString())
            CustomVote(plugin, vote, queued, silent)
        }
    }

    init
    {
        val delay = plugin.config.getNumber(Setting.VOTE_DELAY.path)
        TaskTimer.delay(plugin, delay * 20L) {
            plugin.launch {
                forwardVote()
            }
        }.run()
    }
}